org/qmckl_sherman_morrison_woodbury.org

```diff
@@ -0,0 +1,1775 @@
+#+TITLE: Sherman-Morrison-Woodbury
+#+SETUPFILE: ../tools/theme.setup
+#+INCLUDE: ../tools/lib.org
+
+Low- and high-level functions that use the Sherman-Morrison and
+ Woodbury matrix inversion formulas to update the inverse of a
+ non-singualr matrix
+ 
+* Headers
+  #+begin_src elisp :noexport :results none :exports none
+(org-babel-lob-ingest "../tools/lib.org")
+#+end_src
+
+  #+begin_src c :comments link :tangle (eval c_test) :noweb yes
+#include "qmckl.h"
+#include "assert.h"
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+#include <math.h>
+
+int main() {
+  qmckl_context context;
+  context = qmckl_context_create();
+
+  qmckl_exit_code rc;
+  #+end_src
+
+  
+* Helper Functions
+  
+Helper functions that are used by the Sherman-Morrison-Woodbury kernels. 
+These functions can only be used internally by higher level functions.
+
+** ~qmckl_slagel_splitting~
+   :PROPERTIES:
+   :Name:     qmckl_slagel_splitting
+   :CRetType: double
+   :FRetType: double precision
+   :END:
+
+   ~qmckl_slagel_splitting~ is used internally to apply a list of rank-1 updates while splitting an update if necessary.
+   In case of a split it applies the first half of the update while putting the second half in waiting queue to be applied at the end.
+  
+   For a given update $u_j$ we define a threshold value $\epsilon_j$, which is the minimum value of 
+   $1+v_j^TS^{-1}u_j$ for a non-singular matrix $S$. If $1+v_j^TS^{-1}u_j \geq \epsilon_j$, 
+   the update is applied as usual. Otherwise, $u_j$ will be redefined as $\frac{u_j}{2}$, and the other half 
+   (to be applied at the end) will be defined using vectors $\frac{u_{j'}}{2}$ and $v_{j'}^T=v_{j'}^T$.
+   
+   #+NAME: qmckl_slagel_splitting_args
+   | uint64_t | Dim                            | in    | Leading dimension of Slater_inv                               |
+   | uint64_t | N_updates                      | in    | Number of rank-1 updates to be applied to Slater_inv          |
+   | double   | Updates[N_updates*Dim]         | in    | Array containing the rank-1 updates                           |
+   | uint64_t | Updates_index[N_updates]       | in    | Array containing positions of the rank-1 updates              |
+   | double   | breakdown                      | in    | Break-down parameter on which to fail or not                  |
+   | double   | Slater_inv[Dim*Dim]            | inout | Array containing the inverse Slater-matrix                    |
+   | double   | later_updates[Dim * N_updates] | inout | Array containing the split updates for later                  |
+   | uint64_t | later_index[N_updates]         | inout | Array containing the positions of the split updates for later |
+   | uint64_t | later                          | inout | Number of split updates for later                             |
+  
+
+*** Requirements
+
+    - ~Dim >= 2~
```

Comment:
Expres in terms of element, not in bytes. Also Updates is not multiplied by N_updates! 
 ** Change for all following kernels.